File: class11/python/GQS1101-ChecklistInspecao.py
#!/usr/bin/env python3
"""
GQS1101 - Checklist de Inspeção de Código.
Ferramenta interativa para revisão de código.
Uso: python3 GQS1101-ChecklistInspecao.py
"""

CHECKLIST = {
    "Lógica": [
        "Condições de borda tratadas?",
        "Laços têm condição de parada?",
        "Recursão tem caso base?",
    ],
    "Dados": [
        "Variáveis foram inicializadas?",
        "Parâmetros de métodos são validados?",
        "Null pointers são evitados?",
    ],
    "Tratamento de Erros": [
        "Exceções não foram silenciadas?",
        "Recursos são fechados adequadamente?",
        "Mensagens de erro são informativas?",
    ],
    "Legibilidade": [
        "Nomes são significativos?",
        "Métodos são coesos (uma responsabilidade)?",
        "Comentários são necessários e claros?",
    ],
    "Padrões": [
        "Convenções de codificação seguidas?",
        "Formatação consistente?",
        "Código morto foi removido?",
    ],
}


def executar_checklist():
    print("=" * 60)
    print("  GQS1101 - CHECKLIST DE INSPEÇÃO DE CÓDIGO")
    print("=" * 60)

    total = 0
    aprovados = 0
    falhas = []

    for categoria, itens in CHECKLIST.items():
        print(f"\n--- {categoria} ---")
        for item in itens:
            total += 1
            while True:
                resp = input(f"  {item} (s/n): ").strip().lower()
                if resp in ("s", "n"):
                    if resp == "s":
                        aprovados += 1
                    else:
                        falhas.append(f"{categoria}: {item}")
                    break
                print("    Responda 's' ou 'n'.")

    print("\n" + "=" * 60)
    print("  RESULTADO DA INSPEÇÃO")
    print("=" * 60)
    print(f"  Itens verificados: {total}")
    print(f"  Aprovados: {aprovados}/{total} ({aprovados/total*100:.0f}%)")
    if falhas:
        print(f"  Falhas ({len(falhas)}):")
        for f in falhas:
            print(f"    ❌ {f}")
    print("=" * 60)


if __name__ == "__main__":
    executar_checklist()

File: class11/java/GQS1101-ChecklistInspecao.java
/**
 * GQS1101 - Checklist de Inspeção de Código.
 * Ferramenta interativa para revisão de código, organizada por categorias.
 *
 * Compilação e execução:
 *   javac GQS1101-ChecklistInspecao.java
 *   java -cp . GQS1101_ChecklistInspecao
 */
import java.util.*;

class GQS1101_ChecklistInspecao {

    /** Checklist de inspeção organizado por categoria, na ordem de exibição. */
    static LinkedHashMap<String, List<String>> criarChecklist() {
        LinkedHashMap<String, List<String>> checklist = new LinkedHashMap<>();
        checklist.put("Lógica", List.of(
            "Condições de borda tratadas?",
            "Laços têm condição de parada?",
            "Recursão tem caso base?"
        ));
        checklist.put("Dados", List.of(
            "Variáveis foram inicializadas?",
            "Parâmetros de métodos são validados?",
            "Null pointers são evitados?"
        ));
        checklist.put("Tratamento de Erros", List.of(
            "Exceções não foram silenciadas?",
            "Recursos são fechados adequadamente?",
            "Mensagens de erro são informativas?"
        ));
        checklist.put("Legibilidade", List.of(
            "Nomes são significativos?",
            "Métodos são coesos (uma responsabilidade)?",
            "Comentários são necessários e claros?"
        ));
        checklist.put("Padrões", List.of(
            "Convenções de codificação seguidas?",
            "Formatação consistente?",
            "Código morto foi removido?"
        ));
        return checklist;
    }

    static void executarChecklist(Scanner sc) {
        System.out.println("=".repeat(60));
        System.out.println("  GQS1101 - CHECKLIST DE INSPEÇÃO DE CÓDIGO");
        System.out.println("=".repeat(60));

        LinkedHashMap<String, List<String>> checklist = criarChecklist();
        int total = 0;
        int aprovados = 0;
        List<String> falhas = new ArrayList<>();

        for (Map.Entry<String, List<String>> entrada : checklist.entrySet()) {
            String categoria = entrada.getKey();
            System.out.printf("%n--- %s ---%n", categoria);
            for (String item : entrada.getValue()) {
                total++;
                while (true) {
                    System.out.printf("  %s (s/n): ", item);
                    String resp = sc.nextLine().strip().toLowerCase();
                    if (resp.equals("s") || resp.equals("n")) {
                        if (resp.equals("s")) {
                            aprovados++;
                        } else {
                            falhas.add(categoria + ": " + item);
                        }
                        break;
                    }
                    System.out.println("    Responda 's' ou 'n'.");
                }
            }
        }

        System.out.println("\n" + "=".repeat(60));
        System.out.println("  RESULTADO DA INSPEÇÃO");
        System.out.println("=".repeat(60));
        System.out.printf("  Itens verificados: %d%n", total);
        System.out.printf("  Aprovados: %d/%d (%.0f%%)%n", aprovados, total, aprovados * 100.0 / total);
        if (!falhas.isEmpty()) {
            System.out.printf("  Falhas (%d):%n", falhas.size());
            for (String f : falhas) {
                System.out.printf("    ❌ %s%n", f);
            }
        }
        System.out.println("=".repeat(60));
    }

    public static void main(String[] args) {
        try (Scanner sc = new Scanner(System.in)) {
            executarChecklist(sc);
        }
    }
}
